File: apps/ios/Anky/Core/WriteBeforeScroll/WriteBeforeScrollEventLogStore.swift
import Foundation

enum WriteBeforeScrollEventName: String, Codable, CaseIterable {
    case shieldShown = "shield_shown"
    case shieldRendered = "shield_rendered"
    case shieldActionTapped = "shield_action_tapped"
    case shieldPrimaryTapped = "shield_primary_tapped"
    case ankyOpenedFromShieldPendingState = "anky_opened_from_shield_pending_state"
    case writingStarted = "writing_started"
    case sentenceUnlockAvailable = "sentence_unlock_available"
    case presenceUnlockAvailable = "presence_unlock_available"
    case ankyUnlockAvailable = "anky_unlock_available"
    case unlockTapped = "unlock_tapped"
    case continuedWritingAfterUnlockAvailable = "continued_writing_after_unlock_available"
    case unlockGranted = "unlock_granted"
    case relockScheduled = "relock_scheduled"
    case relockApplied = "relock_applied"
    case relockFailed = "relock_failed"
    case bridgeModeDirectOpen = "bridge_mode_direct_open"
    case bridgeModeNotification = "bridge_mode_notification"
    case directOpenRequested = "direct_open_requested"
    case notificationScheduled = "notification_scheduled"
    case notificationResendTapped = "notification_resend_tapped"
    case notificationPermissionMissing = "notification_permission_missing"
    case notificationTapped = "notification_tapped"
    case appOpenedWithPendingWBSIntent = "app_opened_with_pending_wbs_intent"
    case routedToWBSFromShield = "routed_to_wbs_from_shield"
    case pendingIntentConsumed = "pending_intent_consumed"
    case emergencyUnlockTapped = "emergency_unlock_tapped"
    case onboardingTargetSet = "onboarding_target_set"
    case targetChanged = "target_changed"
    case quickPassUsed = "quick_pass_used"
    case quickPassExhaustedShown = "quick_pass_exhausted_shown"
    case dailyTargetReached = "daily_target_reached"
    case sessionOvershoot = "session_overshoot"
    case screenTimeAuthorizationRequested = "screen_time_authorization_requested"
    case screenTimeAuthorizationGranted = "screen_time_authorization_granted"
    case screenTimeAuthorizationDenied = "screen_time_authorization_denied"
    case appSelectionSaved = "app_selection_saved"
    case shieldApplied = "shield_applied"
    case shieldCleared = "shield_cleared"
    case wbsSessionSealed = "wbs_session_sealed"
}

struct WriteBeforeScrollEvent: Codable, Equatable, Identifiable {
    let id: UUID
    let name: WriteBeforeScrollEventName
    let timestamp: Date
    let sessionID: UUID?
    let tierRawValue: String?
    let message: String?
    let metadata: [String: String]

    init(
        id: UUID = UUID(),
        name: WriteBeforeScrollEventName,
        timestamp: Date = Date(),
        sessionID: UUID? = nil,
        tierRawValue: String? = nil,
        message: String? = nil,
        metadata: [String: String] = [:]
    ) {
        self.id = id
        self.name = name
        self.timestamp = timestamp
        self.sessionID = sessionID
        self.tierRawValue = tierRawValue
        self.message = message
        self.metadata = metadata
    }
}

struct WriteBeforeScrollEventLogStore {
    private let defaults: UserDefaults
    private let key = "writeBeforeScroll.eventLog.v1"
    private let maxStoredEvents: Int

    init(
        defaults: UserDefaults = AnkyAppGroupStorage.userDefaults(),
        maxStoredEvents: Int = 300
    ) {
        self.defaults = defaults
        self.maxStoredEvents = maxStoredEvents
    }

    func load() -> [WriteBeforeScrollEvent] {
        guard let data = defaults.data(forKey: key),
              let events = try? JSONDecoder().decode([WriteBeforeScrollEvent].self, from: data) else {
            return []
        }
        return events
    }

    func recent(limit: Int = 20) -> [WriteBeforeScrollEvent] {
        Array(load().suffix(limit).reversed())
    }

    func append(
        _ name: WriteBeforeScrollEventName,
        at timestamp: Date = Date(),
        sessionID: UUID? = nil,
        tierRawValue: String? = nil,
        message: String? = nil,
        metadata: [String: String] = [:]
    ) {
        append(
            WriteBeforeScrollEvent(
                name: name,
                timestamp: timestamp,
                sessionID: sessionID,
                tierRawValue: tierRawValue,
                message: message,
                metadata: metadata
            )
        )
    }

    func append(_ event: WriteBeforeScrollEvent) {
        var events = load()
        events.append(event)
        if events.count > maxStoredEvents {
            events = Array(events.suffix(maxStoredEvents))
        }
        guard let data = try? JSONEncoder().encode(events) else {
            return
        }
        defaults.set(data, forKey: key)
    }

    func clear() {
        defaults.removeObject(forKey: key)
    }
}

File: apps/android/app/src/main/java/inc/anky/android/core/gate/WriteBeforeScrollEventLogStore.kt
package inc.anky.android.core.gate

import android.content.SharedPreferences
import java.time.Instant
import java.util.UUID
import org.json.JSONArray
import org.json.JSONObject

enum class WriteBeforeScrollEventName(val rawValue: String) {
    ShieldShown("shield_shown"),
    ShieldRendered("shield_rendered"),
    ShieldActionTapped("shield_action_tapped"),
    ShieldPrimaryTapped("shield_primary_tapped"),
    AnkyOpenedFromShieldPendingState("anky_opened_from_shield_pending_state"),
    WritingStarted("writing_started"),
    SentenceUnlockAvailable("sentence_unlock_available"),
    PresenceUnlockAvailable("presence_unlock_available"),
    AnkyUnlockAvailable("anky_unlock_available"),
    UnlockTapped("unlock_tapped"),
    ContinuedWritingAfterUnlockAvailable("continued_writing_after_unlock_available"),
    UnlockGranted("unlock_granted"),
    RelockScheduled("relock_scheduled"),
    RelockApplied("relock_applied"),
    RelockFailed("relock_failed"),
    BridgeModeDirectOpen("bridge_mode_direct_open"),
    BridgeModeNotification("bridge_mode_notification"),
    DirectOpenRequested("direct_open_requested"),
    NotificationScheduled("notification_scheduled"),
    NotificationResendTapped("notification_resend_tapped"),
    NotificationPermissionMissing("notification_permission_missing"),
    NotificationTapped("notification_tapped"),
    AppOpenedWithPendingWBSIntent("app_opened_with_pending_wbs_intent"),
    RoutedToWBSFromShield("routed_to_wbs_from_shield"),
    PendingIntentConsumed("pending_intent_consumed"),
    EmergencyUnlockTapped("emergency_unlock_tapped"),
    OnboardingTargetSet("onboarding_target_set"),
    TargetChanged("target_changed"),
    QuickPassUsed("quick_pass_used"),
    QuickPassExhaustedShown("quick_pass_exhausted_shown"),
    DailyTargetReached("daily_target_reached"),
    SessionOvershoot("session_overshoot"),
    ScreenTimeAuthorizationRequested("screen_time_authorization_requested"),
    ScreenTimeAuthorizationGranted("screen_time_authorization_granted"),
    ScreenTimeAuthorizationDenied("screen_time_authorization_denied"),
    AppSelectionSaved("app_selection_saved"),
    ShieldApplied("shield_applied"),
    ShieldCleared("shield_cleared"),
    WbsSessionSealed("wbs_session_sealed"),
    ;

    companion object {
        fun fromRawValue(rawValue: String?): WriteBeforeScrollEventName? =
            entries.firstOrNull { it.rawValue == rawValue }
    }
}

data class WriteBeforeScrollEvent(
    val id: UUID = UUID.randomUUID(),
    val name: WriteBeforeScrollEventName,
    val timestamp: Instant = Instant.now(),
    val sessionId: UUID? = null,
    val tierRawValue: String? = null,
    val message: String? = null,
    val metadata: Map<String, String> = emptyMap(),
)

class WriteBeforeScrollEventLogStore(
    private val preferences: SharedPreferences,
    private val maxStoredEvents: Int = 300,
) {
    fun load(): List<WriteBeforeScrollEvent> {
        val raw = preferences.getString(Key, null) ?: return emptyList()
        return runCatching {
            val array = JSONArray(raw)
            (0 until array.length()).mapNotNull { index ->
                array.getJSONObject(index).toEventOrNull()
            }
        }.getOrDefault(emptyList())
    }

    fun recent(limit: Int = 20): List<WriteBeforeScrollEvent> =
        load().takeLast(limit).reversed()

    fun append(
        name: WriteBeforeScrollEventName,
        at: Instant = Instant.now(),
        sessionId: UUID? = null,
        tierRawValue: String? = null,
        message: String? = null,
        metadata: Map<String, String> = emptyMap(),
    ) {
        append(
            WriteBeforeScrollEvent(
                name = name,
                timestamp = at,
                sessionId = sessionId,
                tierRawValue = tierRawValue,
                message = message,
                metadata = metadata,
            ),
        )
    }

    fun append(event: WriteBeforeScrollEvent) {
        var events = load() + event
        if (events.size > maxStoredEvents) {
            events = events.takeLast(maxStoredEvents)
        }
        val array = JSONArray()
        events.forEach { array.put(it.toJson()) }
        preferences.edit().putString(Key, array.toString()).apply()
    }

    /**
     * Shield events are transition-only, mirroring iOS
     * `ScreenTimeShieldController`: `shield_applied` logs only when the
     * shield goes up from down, `shield_cleared` only when it comes down
     * from up. Reconciles that re-affirm the current state stay silent.
     */
    fun appendShieldTransition(
        wasShieldActive: Boolean,
        isShieldActive: Boolean,
        at: Instant = Instant.now(),
        metadata: Map<String, String> = emptyMap(),
    ) {
        if (!wasShieldActive && isShieldActive) {
            append(WriteBeforeScrollEventName.ShieldApplied, at = at, metadata = metadata)
        } else if (wasShieldActive && !isShieldActive) {
            append(WriteBeforeScrollEventName.ShieldCleared, at = at)
        }
    }

    fun clear() {
        preferences.edit().remove(Key).apply()
    }

    private fun WriteBeforeScrollEvent.toJson(): JSONObject {
        val json = JSONObject()
            .put("id", id.toString())
            .put("name", name.rawValue)
            .put("timestamp", timestamp.toString())
        sessionId?.let { json.put("sessionID", it.toString()) }
        tierRawValue?.let { json.put("tierRawValue", it) }
        message?.let { json.put("message", it) }
        val metadataJson = JSONObject()
        metadata.forEach { (key, value) -> metadataJson.put(key, value) }
        json.put("metadata", metadataJson)
        return json
    }

    private fun JSONObject.toEventOrNull(): WriteBeforeScrollEvent? {
        val name = WriteBeforeScrollEventName.fromRawValue(optString("name")) ?: return null
        val metadataJson = optJSONObject("metadata") ?: JSONObject()
        return WriteBeforeScrollEvent(
            id = runCatching { UUID.fromString(getString("id")) }.getOrElse { UUID.randomUUID() },
            name = name,
            timestamp = Instant.parse(getString("timestamp")),
            sessionId = optString("sessionID", "").takeIf { it.isNotEmpty() }
                ?.let { runCatching { UUID.fromString(it) }.getOrNull() },
            tierRawValue = if (has("tierRawValue")) getString("tierRawValue") else null,
            message = if (has("message")) getString("message") else null,
            metadata = metadataJson.keys().asSequence().associateWith { metadataJson.getString(it) },
        )
    }

    private companion object {
        const val Key = "writeBeforeScroll.eventLog.v1"
    }
}
